DSMR sensor

aequitas opened the pull request:
This is my initial version of an sensor implementation for the DSMR (Dutch smart meter protocol). In its current state it is blocking IO which I intent to fix first and it is dependent on contributions to `dsmr_parser` which need to be released upstream. 

I would like to share this as a WIP and get feedback about if I am on track as this is the first component implementation I've written.

**Example entry for `configuration.yaml` (if applicable):**
```yaml
sensor:
  - name: dsmr
    platform: dsmr
    device: /dev/ttyUSB1
```

**Checklist:**

Todos:
- [x] Add test 
- [x] Non-blocking IO implemented
- [x] `dsmr_parser` contribution merged upstream (https://github.com/ndokter/dsmr_parser/pull/2)

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully.
  - [x] New dependencies have been added to the `REQUIREMENTS` variable.
  - [x] New dependencies are only imported inside functions that use them.
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - ~~New files were added to `.coveragerc`.~~
  - [x] Tests have been added to verify that the new code works.
  - [x] Entity icons
  - [ ] Error handling and recovery
  - [x] Wait for asyncio protocol to be merged upstream in dsmr_parser, update REQUIREMENTS

aequitas committed:
Initial implemenation of DSMR component.

aequitas committed:
Fix linting

aequitas committed:
Remove protocol V2.2 support until merged upstream.

aequitas committed:
Generate requirements using script.

aequitas committed:
Use updated dsmr-parser with protocol 2.2 support.

aequitas committed:
Add tests.

aequitas committed:
Isort and input validation.

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,153 @@
+
+        _LOGGER.info('got new telegram')
+
+        yield from asyncio.sleep(10, loop=self.hass.loop)
```
What's this sleep for?

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,153 @@
+        self._telegram = {}
+
+    @asyncio.coroutine
+    def async_update(self):
```
As far as I can tell, this isn't making use of any asyncio functionality - this probably shouldn't be async as the `update` func should be run in an executor by the core.

lwis replied:
@aequitas if you're planning on making `dsmr-parser` then your current implementation will be appropriate.

aequitas committed:
Add entities for gas and actual meter reading. Error handling. Use Throttle.

fabaff commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,153 @@
+REQUIREMENTS = ['dsmr-parser==0.2']
+
+CONF_DSMR_VERSION = 'dsmr_version'
+DEFAULT_DEVICE = '/dev/ttyUSB0'
```
I wouldn't recommend to do that because this will most likely not work on macOS and Windows.

fabaff replied:
The issue with that default is that it would be hard to debug for a no-Linux user.

fabaff commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,153 @@
+DEFAULT_DSMR_VERSION = '4'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
```
For serial ports we use `port:` for the configuration.

aequitas replied:
Is this kind of stuff documented somewhere? Because I took the zigbee module as an example which uses 'device' so my guess was that was the convention https://github.com/home-assistant/home-assistant/blob/ece58ce78fe44b521122e3f41b79e670010bc905/homeassistant/components/zigbee.py#L30

and port seems to be generally used only in tcp/udp context.

fabaff replied:
No, it's not documented. I'm trying to establish a common base for the configuration and not use `usb_path`, `filename`, and others. It's a connection over a serial port. Even better than `port` would `serial_port` be, distinguish between the network port and the physical port.

aequitas replied:
Is there a reason it is not outlined in a document yet? 

In my opinion if you want to work toward a convention for (for example) naming you outline it first in a document, this should make it easy for (new) contributors to learn about what the conventions are. Now I had to learn this convention through a pullrequest review, as the only other information I could go on is other components.

aequitas replied:
I rename the config to port, let me know if you want me to take initiative on documenting the conventions.

aequitas committed:
Implement non-blocking serial reader.

aequitas committed:
Improve logging.

aequitas commented:
I added the implementation for a non-blocking reader, however it needs upstream contribution (https://github.com/ndokter/dsmr_parser/pull/3) before it can pass the tests. Feel free to comment on the new code. 

Tests still need to be refactored.

lwis commented on .envrc:
```diff
@@ -0,0 +1 @@
+layout python3
```
Should we keep this? Add it to .gitignore if needed.

aequitas replied:
Sorry, this got included by accident.

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,182 @@
+    hass.loop.create_task(dsmr.read_telegrams(queue))
+
+
+class DSMR:
```
Should this be an interface?

aequitas replied:
Can you explain the question a little further?

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,182 @@
+    @property
+    def state(self):
+        """Return the state of the sensor, if available."""
+        return getattr(self._interface.telegram.get(self._obis, {}),
```
This should be cached in an `update` method.

aequitas replied:
The call itself only reads the value of telegram and does not perform any IO or update sequence. What would be the reason to cache the value in this case?

lwis commented on tests/components/sensor/test_dsmr.py:
```diff
@@ -0,0 +1,43 @@
+        ]),
+    }
+
+    # with patch('homeassistant.components.sensor.dsmr.DSMR.read_telegram',
```
Why is this commented?

aequitas replied:
I still need to refactor the test suite after de code has been refactored to non-blocking. This will be addresses (see todo).

aequitas committed:
Merge entities into one, add icons, fix tests for asyncio.

aequitas committed:
Add error logging for serial reader.

aequitas commented:
Refactored the tests for asyncio and added error handling. Also simplified/refactored some other parts of the code. I have all todos I intended for this feature completed. Could you please review again and let me know what needs to be changed before it can be accepted.

balloob commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,221 @@
+    devices += [DSMREntity(name, obis, dsmr) for name, obis in obis_mapping]
+
+    # setup devices
+    yield from hass.loop.create_task(async_add_devices(devices))
```
`async_add_devices` is a coroutine function, no need to wrap it in create_task

aequitas replied:
ofc, I still need to get a little accustomed to the framework and asyncio.

balloob commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,221 @@
+    reader.add_done_callback(handle_error)
+
+    # add task to receive telegrams and update entities
+    hass.loop.create_task(dsmr.read_telegrams(queue))
```
please use `hass.async_add_job` so that we can keep track of it (and be able to use `hass.block_till_done()`)

Also, please cancel this coroutine when HOMEASSISTANT_STOP event is fired.

balloob commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -0,0 +1,221 @@
+        """Return the state of sensor, if available, translate if needed."""
+        from dsmr_parser import obis_references as obis
+
+        value = getattr(self._interface.telegram.get(self._obis, {}),
```
It's unclear to me if these calls are doing I/O. They shouldn't (as we're inside async land)

aequitas replied:
Someone else also commented on this. I will refactor/document this piece a bit to make it more clear it is not  doing I/O.

balloob commented on tests/components/sensor/test_dsmr.py:
```diff
@@ -0,0 +1,56 @@
+
+    # mock queue for injecting DSMR telegram
+    queue = asyncio.Queue(loop=hass.loop)
+    monkeypatch.setattr('asyncio.Queue', lambda: queue)
```
Nice. Didn't know about monkeypatch.

aequitas replied:
There is a lot of cool stuff in pytest. Thanks for going with that framework (and Tox). In my opinion it is the best one out there atm, but it doesn't force you into one style of writing tests as it also supports `unittest`.

aequitas replied:
btw, have you considered using https://github.com/klen/pylama for the project? It combines a lot of the linting tools used in this project into a single consistent interface.

aequitas committed:
Refactoring and documentation.

- refactor asyncio reader task to make sure it stops with HA
- document general principle of this component
- refactor entity reading to be more clear
- remove cruft from split entity implementation

aequitas commented:
@balloob I refactored to address the issues you mentioned. Thanks for pointing out the required structures in the framework. I did not find good examples myself in the other components because there was just to much :). My assumption was hass/asyncio would take care of the stopping of the tasks automatically..

I added some additional overall documentation at the top of the module.

Let me know if there is anything else that needs to be done.

aequitas committed:
Use `port` configuration key.

aequitas commented:
Strange, this build failed for py34 but another travis build on the same commit succeeded (https://travis-ci.org/aequitas/home-assistant/builds/175541781). This one might succeed if retried again. Error/failure is not on my testsuite btw. 

Is it common for these kind of test instabilities to pop up in this framework? Or could it be a bug/race-condition introduced by my code?

aequitas committed:
DSMR V2.2 seems to conflict in explaining which tariff is high and low.

http://www.netbeheernederland.nl/themas/hotspot/hotspot-documenten/?dossierid=11010056&title=Slimme%20meter&onderdeel=Documenten
> DSMR v2.2 Final P1
>> 6.1: table vs table note

    Meter Reading electricity delivered to client normal tariff) in 0,01 kWh - 1-0:1.8.1.255
    Meter Reading electricity delivered to client (low tariff) in 0,01 kWh - 1-0:1.8.2.255

    Note: Tariff code 1 is used for low tariff and tariff code 2 is used for normal tariff.

aequitas commented:
Currently this components is being tested by a few persons in this thread: https://community.home-assistant.io/t/support-for-reading-dutch-smart-meter-electricity-gas-p1-port/1676

Some people have issues regarding serial port settings which can differ depending on the meter type and usb hardware used. I am working on finding a solution for this.

henkvdt commented:
I might be wrong, but I think the kWh and kW units of measurement are switched: The current usage shows kW, but this should be kWh (usage per hour). The meter reading shows kWh, but this should be kW (absolute number)

aequitas commented:
@henkvdt the `kW` and `kWh` units are taken directly from the DSMR telegram. So the telegram would be wrong in that case. 

The current usage shows how much power is being used at this moment in kilowats, the meter reading is how much energy has been consumed. 
http://www.energylens.com/articles/kw-and-kwh

aequitas commented:
Users are reporting metrics stalling when a telegram failed to be parsed. Will be looking into this.

aequitas committed:
Refactor to use asyncio.Protocol instead of loop+queue.

aequitas commented:
I refactored the component a little bit to remove some complexity. I learned Protocols is the way to go in asyncio, so that is the implementation I went for this time.

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -26,18 +26,21 @@
 
-REQUIREMENTS = ['dsmr-parser==0.3']
+REQUIREMENTS = [
+    'https://github.com/aequitas/dsmr_parser/archive/async_protocol.zip'
```
Do you know if this will be merged upstream? Or are you planning on maintaining an async fork?

This also needs a version at the end `...zip#dsmr-parser==0.4` for example.

aequitas replied:
Merged upstream, just waiting for new version to be deployed: https://github.com/ndokter/dsmr_parser/pull/7

lwis commented on homeassistant/components/sensor/dsmr.py:
```diff
@@ -84,122 +88,48 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+        for device in devices:
+            device.telegram = telegram
+            hass.async_add_job(device.async_update_ha_state)
+            # hass.loop.create_task(device.async_update_ha_state())
```
What's this for?

aequitas replied:
Nicely spot, left that in there by accident.

aequitas replied:
May I suggest adding: https://github.com/aequitas/pytest-eradicate to the project?

aequitas committed:
Fix requirements

aequitas committed:
Close transport when HA stops.

aequitas committed:
Cleanup.

aequitas committed:
Include as dependency for testing (until merged upstream.)

aequitas committed:
Fix style.

balloob committed:
Update setup.cfg

balloob commented:
Great 🐬

aequitas commented:
@balloob thanks. 

I had some todo's still unchecked in the PR description. What is the recommended way to indicate a PR should not yet be merged? I though leaving todos unchecked would be a good indicator for that.

Merge should not have to be reverted I am pushing the update in a minute and the error handling seems to only affect one of the test users atm with which I am in debug session on the community forum.

lwis commented:
@aequitas if you create a new PR and tag me I'll review and merge if it all looks good. 

If you don't want something to be merged and you're just looking for feedback, it's best to put "WIP" in the title.

aequitas commented:
@lwis ok, this one was a little in between because at first I considered it ready for merge but then some users had feedback already. I will keep with the WIP convention next time.

balloob commented:
Sorry, missed the to do checkbox. WIP is the best indicator.
